src/main/java/rx/internal/operators/OnSubscribeRedo.java

```diff
@@ -235,10 +235,11 @@ public void onError(Throwable e) {
                     @Override
                     public void onNext(T v) {
                         if (!done) {
-                            if (consumerCapacity.get() != Long.MAX_VALUE) {
-                                consumerCapacity.decrementAndGet();
-                            }
                             child.onNext(v);
+                            long cap = consumerCapacity.get();
+                            if (cap != Long.MAX_VALUE) {
+                                consumerCapacity.compareAndSet(cap, cap -1);
```

Comment:
yep of course it would need a loop if we went for the CAS option. I can chuck one in, your call.